keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,2144 @@
+    // - The Cordon or Drain stage can be skipped by setting the stage to Complete.
+    // - The NodeMaintenance object is moved to the Complete stage on deletion unless the Idle stage has been set.
+    //
+    // The default value is Idle.
+    Stage NodeMaintenanceStage
+
+    // DrainPlan is executed from the first entry to the last entry during the Drain stage.
+    // DrainPlanEntry podType fields should be in the following order:
+    // nil -> DaemonSet -> Static
+    // DrainPlanEntry priority fields should be in ascending order for each podType.
+    // If the priority and podType are the same, concrete selectors are executed first.
+    //
+    // The following entries are injected into the drainPlan on the NodeMaintenance admission:
+    // - podPriority: 1000000000 // highest priority for user defined priority classes
+    //   podType: "Default"
+    // - podPriority: 2000000000 // system-cluster-critical priority class
+    //   podType: "Default"
+    // - podPriority: 2000001000 // system-node-critical priority class
+    //   podType: "Default"
+    // - podPriority: 2147483647 // maximum value
+    //   podType: "Default"
+    // - podPriority: 1000000000 // highest priority for user defined priority classes
+    //   podType: "DaemonSet"
+    // - podPriority: 2000000000 // system-cluster-critical priority class
+    //   podType: "DaemonSet"
+    // - podPriority: 2000001000 // system-node-critical priority class
+    //   podType: "DaemonSet"
+    // - podPriority: 2147483647 // maximum value
+    //   podType: "DaemonSet"
+    // - podPriority: 1000000000 // highest priority for user defined priority classes
+    //   podType: "Static"
+    // - podPriority: 2000000000 // system-cluster-critical priority class
+    //   podType: "Static"
+    // - podPriority: 2000001000 // system-node-critical priority class
+    //   podType: "Static"
+    // - podPriority: 2147483647 // maximum value
+    //   podType: "Static"
+    //
+    // Duplicate entries are not allowed.
+    // This field is immutable.
+    DrainPlan []DrainPlanEntry
+
+    // Reason for the maintenance.
+    Reason string
+}
+
+const (
+// Default selects all pods except DaemonSet and Static pods.	
+Default PodType = "Default"
+// DaemonSet selects DaemonSet pods.
+DaemonSet PodType = "DaemonSet"
+// Static selects static pods.
+Static PodType = "Static"
+)
+
+type DrainPlanEntry struct {
+    // PodSelector selects pods according to their labels.
+	// This can help to select which pods of the same priority should be evacuated first.
+    // +optional
+    PodSelector *metav1.LabelSelector
+    // PodPriority specifies a pod priority.
+    // Pods with a priority less or equal to this value are selected.
+    PodPriority int32
+    // PodType selects pods according to the pod type:
+    // - Default selects all pods except DaemonSet and Static pods.	
+    // - DaemonSet selects DaemonSet pods.
+    // - Static selects static pods.
+    PodType PodType
+}
```

Comment:
```suggestion
type DrainPlanEntry struct {
    // PodSelector selects pods according to their labels.
    // This can help to select which pods of the same priority should be evacuated first.
    // +optional
    PodSelector *metav1.LabelSelector
    // PodPriority specifies a pod priority.
    // Pods with a priority less or equal to this value are selected.
    PodPriority int32
    // PodType selects pods according to the pod type:
    // - Default selects all pods except DaemonSet and Static pods.	
    // - DaemonSet selects DaemonSet pods.
    // - Static selects static pods.
    PodType PodType
}
```